src/plugins/newsfeed/public/lib/api.ts

```diff
@@ -0,0 +1,187 @@
+ * ownership. Elasticsearch B.V. licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+import * as Rx from 'rxjs';
+import moment from 'moment';
+import { i18n } from '@kbn/i18n';
+import { catchError, filter, mergeMap, tap } from 'rxjs/operators';
+import { HttpServiceBase } from 'src/core/public';
+import { NEWSFEED_LAST_FETCH_STORAGE_KEY, NEWSFEED_HASH_SET_STORAGE_KEY } from '../../constants';
+import { NewsfeedPluginInjectedConfig, ApiItem, NewsfeedItem, FetchResult } from '../../types';
+
+type ApiConfig = NewsfeedPluginInjectedConfig['newsfeed']['service'];
+
+export class NewsfeedApiDriver {
+  private readonly loadedTime = moment();
+
+  constructor(
+    private readonly kibanaVersion: string,
+    private readonly userLanguage: string,
+    private readonly fetchInterval: number
+  ) {}
+
+  shouldFetch(): boolean {
+    const lastFetch: string | null = sessionStorage.getItem(NEWSFEED_LAST_FETCH_STORAGE_KEY);
+    if (lastFetch == null) {
+      return true;
+    }
+    const last = moment(lastFetch, 'x'); // parse as unix ms timestamp
+
+    // does the last fetch time precede the time that the page was loaded?
+    if (this.loadedTime.diff(last) > 0) {
+      return true;
+    }
+
+    const now = moment();
+    const duration = moment.duration(now.diff(last));
+
+    return duration.asMilliseconds() > this.fetchInterval;
+  }
+
+  updateLastFetch() {
+    sessionStorage.setItem(NEWSFEED_LAST_FETCH_STORAGE_KEY, Date.now().toString());
+  }
+
+  updateHashes(items: NewsfeedItem[]): { previous: string[]; current: string[] } {
+    // replace localStorage hashes with new hashes
+    const stored: string | null = localStorage.getItem(NEWSFEED_HASH_SET_STORAGE_KEY);
+    let old: string[] = [];
+    if (stored != null) {
+      old = stored.split(',');
+    }
+
+    const newHashes = items.map(i => i.hash);
+    const updatedHashes = [...new Set(old.concat(newHashes))];
+    localStorage.setItem(NEWSFEED_HASH_SET_STORAGE_KEY, updatedHashes.join(','));
+
+    return { previous: old, current: updatedHashes };
+  }
+
+  fetchNewsfeedItems(http: HttpServiceBase, config: ApiConfig): Rx.Observable<FetchResult> {
+    const urlPath = config.pathTemplate.replace('{VERSION}', this.kibanaVersion);
+    const fullUrl = config.urlRoot + urlPath;
+
+    return Rx.from(
+      http
+        .fetch(fullUrl, {
+          method: 'GET',
+        })
+        .then(({ items }) => this.modelItems(items))
+    );
```

Comment:
If the fetch fails we might get console errors over the unhandled rejected promise (from fetch).
I'm not sure if the UI gets stuck in a loading state or not, but we should probably handle the failure state.

Reply:
All errors that raised in fetchNewsfeedItems, handled under the pipe catchError, so this failing doesn't stop the Rx.timer from executing next request and news from appearing in UI: 
`return Rx.timer(0, config.mainInterval).pipe(
    filter(() => driver.shouldFetch()),
    mergeMap(() =>
      driver.fetchNewsfeedItems(http, config.service).pipe(`
       

>  catchError(err => {

  `        window.console.error(err);
          return Rx.of({
            error: err,
            kibanaVersion,
            hasNew: false,
            feedItems: [],
          });
        })
      )
    ),
    tap(() => driver.updateLastFetch())
  );`

We have a console error in this case as well:

> backend.js:6 Error: Failed to fetch
>     at _callee6$ (http_setup.ts:397)
>     at tryCatch (runtime.js:45)
>     at Generator.invoke [as _invoke] (runtime.js:271)
>     at Generator.prototype.<computed> [as throw] (runtime.js:97)
>     at asyncGeneratorStep (http_setup.ts:40)
>     at _throw (http_setup.ts:42)

Reply:
++